chameleon-core/src/main/java/org/hibernate/omm/jdbc/MongoPreparedStatement.java

```diff
@@ -332,4 +301,37 @@ public void setCharacterStream(final int parameterIndex, @Nullable final Reader
         throw new NotYetImplementedException();
     }
 
+    @Override
+    protected void replaceParameterMarkers(final BsonDocument command) {
+        recursivelyReplaceParameterMarkers(command, 1);
+    }
+
+    private int recursivelyReplaceParameterMarkers(BsonDocument command, int curParameterIndex) {
+        for (Map.Entry<String, BsonValue> entry : command.entrySet()) {
+            if (entry.getValue().getBsonType() == BsonType.UNDEFINED) {
+                entry.setValue(parameters.get(curParameterIndex));
+                curParameterIndex++;
+            } else if (entry.getValue().isDocument()) {
+                curParameterIndex = recursivelyReplaceParameterMarkers(entry.getValue().asDocument(), curParameterIndex);
+            } else if (entry.getValue().isArray()) {
+                curParameterIndex = recursivelyReplaceParameterMarkers(entry.getValue().asArray(), curParameterIndex);
+            }
+        }
+        return curParameterIndex;
+    }
+
+    private int recursivelyReplaceParameterMarkers(BsonArray array, int curParameterIndex) {
+        for (int i = 0; i < array.size(); i++) {
+            BsonValue value = array.get(i);
+            if (value.getBsonType() == BsonType.UNDEFINED) {
+                array.set(i, parameters.get(curParameterIndex));
+                curParameterIndex++;
+            } else if (value.isDocument()) {
+                curParameterIndex = recursivelyReplaceParameterMarkers(value.asDocument(), curParameterIndex);
+            } else if (value.isArray()) {
+                curParameterIndex = recursivelyReplaceParameterMarkers(value.asArray(), curParameterIndex);
+            }
+        }
+        return curParameterIndex;
+    }
```

Comment:
The above two methods still share some code duplication. Ideally we could move the common stuff into some commmon method, ending up with emphasizing the salient features or difference between them (mainly different value getter and modifier logics; I tried hard to combine both into one elegant method but failed).

Below is some minor improvement I could think of. I reused the 'MutableInteger' class in Hiberante to make code cleaner and arguably more readable:

```
    @Override
    protected void replaceParameterMarkers(final BsonDocument command) {
        recursivelyReplaceParameterMarkers(command, new MutableInteger(1));
    }

    private void recursivelyReplaceParameterMarkers(BsonDocument command, MutableInteger curParameterIndex) {
        for (Map.Entry<String, BsonValue> entry : command.entrySet()) {
            if (entry.getValue().getBsonType() == BsonType.UNDEFINED) {
                entry.setValue(parameters.get(curParameterIndex.getAndIncrement()));
            } else {
                recursivelyReplaceParameterMarkers(entry.getValue(), curParameterIndex);
            }
        }
    }

    private void recursivelyReplaceParameterMarkers(BsonArray array, MutableInteger curParameterIndex) {
        for (int i = 0; i < array.size(); i++) {
            BsonValue value = array.get(i);
            if (value.getBsonType() == BsonType.UNDEFINED) {
                array.set(i, parameters.get(curParameterIndex.getAndIncrement()));
            } else {
                recursivelyReplaceParameterMarkers(value, curParameterIndex);
            }
        }
    }

    private void recursivelyReplaceParameterMarkers(BsonValue value, MutableInteger curParameterIndex) {
        if (value.isDocument()) {
            recursivelyReplaceParameterMarkers(value.asDocument(), curParameterIndex);
        } else if (value.isArray()) {
            recursivelyReplaceParameterMarkers(value.asArray(), curParameterIndex);
        }
    }
```
However, the code duplication is not serious, so I am perfectly fine with your current code. Anyway, this is mainly for POC purpose and we could leave code quality issue later.